piff/writers.py

```diff
@@ -0,0 +1,257 @@
+        combining it with any base names added if this `Writer` was created by
+        the `nested` method.
+        """
+        raise NotImplementedError()
+
+
+class FitsWriter(Writer):
+    """A `Writer` implementation that writes to multiple FITS HDUs.
+
+    This reader is intended to writes files that would be readable Piff before
+    the `Reader` and `Writer` abstractions were added.
+
+    `FitsWriter.nested` yields a writer that writes to the same file and
+    prepends all names with its base name to form the EXTNAME, concatenating
+    them with ``_``.
+
+    :param fits:        An already-open `fitsio.FITS` object.
+    :param base_name:   Base name to prepend to all object names, or `None`.
+    :param header:      Fields to be added to all FITS extension headers.
+    """
+
+    def __init__(self, fits, base_name, header):
+        self._fits = fits
+        self._base_name = base_name
+        self._header = header
+
+    @classmethod
+    @contextmanager
+    def _open(cls, file_name):
+        """Return a context manager that opens the given FITS file and yields
+        a `FitsWriter` object.
+
+        :param file_name:   Name of the file to read.
+
+        :returns:  A context manager that yields a `FitsWriter` instance.
+
+        This also adds an empty primary HDU and sets up the returned writer
+        to add the Piff version to all HDUs (including the primary).
+        """
+        from . import __version__ as piff_version
+
+        header = {"piff_version": piff_version}
+        with fitsio.FITS(file_name, "rw", clobber=True) as f:
+            if len(f) == 1:
+                f.write(data=None, header=header)
+            yield cls(f, base_name=None, header=header.copy())
+
+    def write_struct(self, name, struct):
+        # Docstring inherited.
+        cols = []
+        dtypes = []
+        for key, value in struct.items():
+            # Don't add values that are None to the table.
+            if value is None:
+                continue
+            dt = make_dtype(key, value)
+            value = adjust_value(value, dt)
+            cols.append([value])
+            dtypes.append(dt)
+        table = np.array(list(zip(*cols)), dtype=dtypes)
+        return self.write_table(name, table)
+
+    def write_table(self, name, array, metadata=None):
+        # Docstring inherited.
+        if metadata:
+            header = self._header.copy()
+            header.update(metadata)
+        else:
+            header = self._header
+        self._fits.write_table(array, extname=self.get_full_name(name), header=header)
+
+    def write_array(self, name, array, metadata=None):
+        # Docstring inherited.
+        if metadata:
+            header = self._header.copy()
+            header.update(metadata)
+        else:
+            header = self._header
+        self._fits.write_image(
+            array, extname=self.get_full_name(name), header=self._header
+        )
```

Comment:
This method seems not to be used?  We don't have any PSF types that write images, so I think for now write_table is all we need.  Although I guess it's not crazy to imagine we might want the option to write images someday.  But it will need some kind of unit test until then, since right now this function doesn't seem to be called by anything.

Reply:
Huh, I hadn't realized that.  I think it'd probably be best to just remove it, then, and if we need it someday that will clarify what it should actually look like.